### ISSUE TITLE
Remove(<lock/5a53798b51>) returned error

### ISSUE THREAD

**Issue by bcHelix** [restic/restic#2452]
Output of `restic version`
--------------------------
restic 0.9.4 compiled with go1.11.6 on linux/amd64

How did you run restic exactly?
-------------------------------
Via a script that passes arguments to restic.
```
restic backup --repo b2:${bucket} $(awk 'BEGIN{ ORS=" " }{ print }' "${includeList}") --verbose=2 \
 --password-file=<(decrypt "${bucket}" | sed -n 3p) -o b2.connections=100
```

What backend/server/service did you use to store the repository?
----------------------------------------------------------------

Backblaze B2

Expected behavior
-----------------
Not receiving lock errors.

Actual behavior
---------------
`Remove(<lock/5a53798b51>) returned error, retrying after 3.588215337s: Delete: b2_download_file_by_name: 404: bucket [bucket] does not have file: locks/5a53798b511f2fe2e7b31d98bf507834270dad5974b365f3dd9db8a19a70b3be`

`unable to refresh lock: Delete: b2_download_file_by_name: 404: bucket [bucket] does not have file: locks/5a53798b511f2fe2e7b31d98bf507834270dad5974b365f3dd9db8a19a70b3be`

I am getting the above errors when attempting to run a backup. This is the first time I've seen these errors occur and there are many of them. The errors only occur on one b2 bucket, the other buckets are fine. The script runs an unlock command on the repo before backing up, and the errors are still present after doing an `unlock` and `prune` manually. All the data appears to have been backed up successfully, and this is the first time I've seen this in over a year of using Restic.

Steps to reproduce the behavior
-------------------------------
Run a backup or prune.

Do you have any idea what may have caused this?
-----------------------------------------------
Nope

Do you have an idea how to solve the issue?
-------------------------------------------
Possibly delete the files from the repo and reupload, however due to the large about of data this is not desirable.

Did restic help you or made you happy in any way?
-------------------------------------------------

**Comment by drivenbyentropy** [comment]
I am getting the exact same error messages. Currently half way through a 1.2 TB backup. Will report back once it finishes. 

Of note, my first attempt of backing up my files got killed due to insufficient memory on my QNAP station. I consequently increased the swap space and restarted the backup to my B2 bucket. This is when these error messages started appearing.

**Comment by Kramins** [comment]
I'm currently running on a prune on 1.8TB backup and I can getting the same error. Seems to be one file that it keeps trying to remove.

**Comment by vespian** [comment]
Same here, both during the backup as well as index rebuild.

**Comment by tempname1024** [comment]
Having this problem too. `restic unlock` ran successfully during a backup where the error was present but restic still tries to remove the (non-existent) lock file.

`restic 0.9.6 compiled with go1.13.4 on linux/amd64`

**Comment by NeoLeMarc** [comment]
Same here when trying any operations on a restic repository to sftp running on a synology diskstation. Restic Version is 0.8.3.

**Comment by DrMurx** [comment]
Manually uploading an empty file to the `locks/` folder (might need creation) of the B2 bucket with the given filename makes `restic unlock` working again.

**Comment by rawtaz** [comment]
@NeoLeMarc You should probably update your restic version, 0.8.3 is really old :)

To everyone this happens to, when this happens it would be good to check the storage and see if the file that restic says isn't there is there or not. It shouldn't be there. If it is, it's possible that there's a race condition between the API connection restic uses and whatever way in oneself is using to check the file's existance. It's of course also possible that there's a bug in restic so it tries to locate the wrong file, but that's probably less likely considering how few reports we get about this.

**Comment by Abukamel** [comment]
I am having the same issues restic `0.9.6` using b2 as a repo and it has lots of files 22 TB of 5 snapshots and I am doing a `forget --prune`, actually restic always fails with me when I am backing up huge amount of files especially obvious when I am backing up a MinIO s3 instance as it always have hundreds of GBs of static media files and pictures.
![image](https://user-images.githubusercontent.com/1246960/79045850-aa315400-7c0d-11ea-907d-974b225c8812.png)
I am really disappointed about this issue for more than a year now and I don't know how to solve it.

I tried `restic check` and `rebuild-index`, almost all the recommendation and nothing worked with me, as long as I am doing  backup for a huge data set, I just get into this error no matter what.

This error was actually happening to me across smaller data set when I was using B2 US region, once they announced EU region, I migrated all my US buckets to EU ones and errors reduced dramatically and is now specific only to my MinIO s3 instances as they have hundreds of GBs of static data as I mentioned.

I wish that there is a solution to this.

**Comment by MichaelEischer** [comment]
@Abukamel The log picture only shows that restic complains about a weird lock file. Do the backup / prune operations actually fail or is restic just spamming the log with these warnings? When do these warnings show up, right at the beginning of the backup run or some time inbetween?

The error messages in your log picture are not related to the `check` and `rebuild-index` command. For problems with lock files you'd rather wan to use the `unlock` command.

> Manually uploading an empty file to the `locks/` folder (might need creation) of the B2 bucket with the given filename makes `restic unlock` working again.

Did you try these steps from @DrMurx ?

P.S.: Please paste log excerpts as text and enclose them in tripple backticks "\`\`\`".

**Comment by MichaelEischer** [comment]
@bcHelix Do you still have the full log of the problematic prune run? It would be interesting to see the _first_ `Remove(<lock/[...]>) return error` line. I'm asking because that would show whether your error is similar to one that showed up in #2473.

**Comment by Abukamel** [comment]
I think that what is happening is that `forget --prune` command takes too much time to finish, for example right now I stopped the daily cron job and run the script directly, it's been 41 hours till now and the prune command is not done yet.

So it keeps trying to remove the old snapshot according to policy for more than one day, then cron job runs again next day and it runs a script that starts with unlock command to the repo and here we start to see errors regarding lock files, so I am not sure how to keep running the daily backup script while applying the policy of old snapshot deletion if I have a big data set like this one (~4TB)
![image](https://user-images.githubusercontent.com/1246960/79119304-569b4380-7d90-11ea-8c58-a9f0afb58256.png)

**Comment by MichaelEischer** [comment]
The problem here is that the lock refresh in restic 0.9.6 and older fails to update the lock file timestamp (fixed in #2391) which later on allows `unlock` to delete an still active lock file.

You _cannot_ run prune and backup in parallel. This is for the time being a necessary restriction to ensure that prune does not remove data used by new snapshots. Automatically running `unlock` is rather dangerous, if you don't ensure that no other restic runs are active at the moment.

**Comment by ProactiveServices** [comment]
> I think that what is happening is that `forget --prune` command takes too much time to finish, for example right now I stopped the daily cron job and run the script directly, it's been 41 hours till now and the prune command is not done yet.
> 
> So it keeps trying to remove the old snapshot according to policy for more than one day, then cron job runs again next day and it runs a script that starts with unlock command to the repo and here we start to see errors regarding lock files, so I am not sure how to keep running the daily backup script while applying the policy of old snapshot deletion if I have a big data set like this one (~4TB)

In my scripts I use my own lock files to decide whether an operation is safe to run. This should also work in your case.
